File: apps/ios/AzriShared/Models/ContractTypes.swift
// AzriShared/Models/ContractTypes.swift
//
// Swift mirror of the wire types declared in @azri/contracts. Kept in
// hand-sync because Swift codegen from Zod is not in our toolchain yet.
//
// ⚠ When @azri/contracts ships a MAJOR bump, this file MUST be updated
//   in the same PR. CI checks the SCHEMA_VERSION constant against the
//   value in packages/contracts/src/version.ts.

import Foundation

public enum SchemaVersion {
    public static let current = "0.2.0"
}

public enum DeviceKind: String, Codable {
    case iosPhone = "ios-phone"
    case iosWatch = "ios-watch"
    case androidPhone = "android-phone"
    case wearOs = "wear-os"
    case tizen
    case whoop
    case web
    case serverTest = "server-test"
    case unknown
}

public enum PhiClassification: String, Codable {
    case publicData = "public"
    case `internal` = "internal"
    case personal
    case health
    case healthSensitive = "health-sensitive"
}

public enum BiosignalKind: String, Codable {
    case heartRate = "heart_rate"
    case heartRateVariability = "heart_rate_variability"
    case restingHeartRate = "resting_heart_rate"
    case oxygenSaturation = "oxygen_saturation"
    case respiratoryRate = "respiratory_rate"
    case skinTemperatureDelta = "skin_temperature_delta"
    case electrodermalActivity = "electrodermal_activity"
    case sleepStage = "sleep_stage"
    case stepCount = "step_count"
    case activeEnergy = "active_energy"
    case fallDetection = "fall_detection"
    case seizureSensor = "seizure_sensor"
}

public enum BiosignalUnit: String, Codable {
    case bpm
    case ms
    case percent
    case breathsPerMin = "breaths_per_min"
    case celsiusDelta = "celsius_delta"
    case microsiemens
    case count
    case kcal
    case score
    case boolean
    case stage
}

public struct EventSource: Codable, Equatable {
    public var deviceKind: DeviceKind
    public var deviceModel: String?
    public var osVersion: String?
    public var appVersion: String?
    public var installId: String?

    public init(
        deviceKind: DeviceKind,
        deviceModel: String? = nil,
        osVersion: String? = nil,
        appVersion: String? = nil,
        installId: String? = nil
    ) {
        self.deviceKind = deviceKind
        self.deviceModel = deviceModel
        self.osVersion = osVersion
        self.appVersion = appVersion
        self.installId = installId
    }
}

public struct BiosignalSample: Codable, Equatable {
    public var at: Date
    public var value: Double
    public var confidence: Double?

    public init(at: Date, value: Double, confidence: Double? = nil) {
        self.at = at
        self.value = value
        self.confidence = confidence
    }
}

/// Wire envelope mirrors @azri/contracts BiosignalBatchEvent.
public struct BiosignalBatchEvent: Codable {
    public var type: String = "biosignal.batch"
    public var eventId: UUID
    public var schemaVersion: String
    public var occurredAt: Date
    public var organizationId: UUID
    public var patientId: UUID
    public var source: EventSource
    public var phiClassification: PhiClassification
    public var kind: BiosignalKind
    public var unit: BiosignalUnit
    public var samples: [BiosignalSample]
    public var samplingHz: Double?

    public init(
        eventId: UUID,
        occurredAt: Date,
        organizationId: UUID,
        patientId: UUID,
        source: EventSource,
        phiClassification: PhiClassification = .healthSensitive,
        kind: BiosignalKind,
        unit: BiosignalUnit,
        samples: [BiosignalSample],
        samplingHz: Double? = nil
    ) {
        self.eventId = eventId
        self.schemaVersion = SchemaVersion.current
        self.occurredAt = occurredAt
        self.organizationId = organizationId
        self.patientId = patientId
        self.source = source
        self.phiClassification = phiClassification
        self.kind = kind
        self.unit = unit
        self.samples = samples
        self.samplingHz = samplingHz
    }
}

/// Wire envelope for AlertRaisedEvent (subset; bodies omitted from
/// scaffold for brevity — fill before the watchOS alert UI ships).
public struct AlertRaisedEventLite: Codable {
    public let type = "alert.raised"
    public var eventId: UUID
    public var schemaVersion: String = SchemaVersion.current
    public var occurredAt: Date
    public var organizationId: UUID
    public var patientId: UUID
    public var source: EventSource
    public var phiClassification: PhiClassification
    public var alertId: UUID
    public var severity: String  // "info" | "warning" | "urgent" | "emergency"
    public var reason: String
    public var channels: [String]
}

File: apps/wear-os/app/src/main/java/ai/azri/wear/ContractTypes.kt
package ai.azri.wear

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

/**
 * Kotlin mirror of @azri/contracts wire types — kept hand-in-hand with the
 * Zod schemas in packages/contracts. When SCHEMA_VERSION bumps MAJOR,
 * update this file in the same PR. CI checks the constant matches the
 * Zod-side version string.
 */
object SchemaVersion { const val current = "0.2.0" }

@Serializable
enum class DeviceKind {
    @SerialName("ios-phone") IOS_PHONE,
    @SerialName("ios-watch") IOS_WATCH,
    @SerialName("android-phone") ANDROID_PHONE,
    @SerialName("wear-os") WEAR_OS,
    @SerialName("tizen") TIZEN,
    @SerialName("whoop") WHOOP,
    @SerialName("web") WEB,
    @SerialName("server-test") SERVER_TEST,
    @SerialName("unknown") UNKNOWN,
}

@Serializable
enum class PhiClassification {
    @SerialName("public") PUBLIC,
    @SerialName("internal") INTERNAL,
    @SerialName("personal") PERSONAL,
    @SerialName("health") HEALTH,
    @SerialName("health-sensitive") HEALTH_SENSITIVE,
}

@Serializable
enum class BiosignalKind {
    @SerialName("heart_rate") HEART_RATE,
    @SerialName("heart_rate_variability") HEART_RATE_VARIABILITY,
    @SerialName("resting_heart_rate") RESTING_HEART_RATE,
    @SerialName("oxygen_saturation") OXYGEN_SATURATION,
    @SerialName("respiratory_rate") RESPIRATORY_RATE,
    @SerialName("step_count") STEP_COUNT,
    @SerialName("active_energy") ACTIVE_ENERGY,
    @SerialName("fall_detection") FALL_DETECTION,
}

@Serializable
enum class BiosignalUnit {
    @SerialName("bpm") BPM,
    @SerialName("ms") MS,
    @SerialName("percent") PERCENT,
    @SerialName("breaths_per_min") BREATHS_PER_MIN,
    @SerialName("count") COUNT,
    @SerialName("kcal") KCAL,
    @SerialName("score") SCORE,
    @SerialName("boolean") BOOLEAN,
}

@Serializable
data class EventSource(
    val deviceKind: DeviceKind,
    val deviceModel: String? = null,
    val osVersion: String? = null,
    val appVersion: String? = null,
    val installId: String? = null,
)

@Serializable
data class BiosignalSample(
    val at: String,                  // ISO-8601 UTC
    val value: Double,
    val confidence: Double? = null,
)

@Serializable
data class BiosignalBatchEvent(
    val type: String = "biosignal.batch",
    val eventId: String,
    val schemaVersion: String = SchemaVersion.current,
    val occurredAt: String,
    val organizationId: String,
    val patientId: String,
    val source: EventSource,
    val phiClassification: PhiClassification = PhiClassification.HEALTH_SENSITIVE,
    val kind: BiosignalKind,
    val unit: BiosignalUnit,
    val samples: List<BiosignalSample>,
    val samplingHz: Double? = null,
)

@Serializable
data class WatchIngestRequest(
    val batchId: String,
    val sentAt: String,
    val events: List<BiosignalBatchEvent>,
)
